New option called `"best"` for `args.save_strategy`.

seanswyi opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Addresses https://github.com/huggingface/transformers/issues/31626.

Adds a new option called `"best"` for `TrainingArguments.save_strategy` which saves the model checkpoint each time a new best performance is achieved.

### Details

1. The previous `_save_checkpoint` method was in charge of not only saving the model checkpoint but also determining the best metric and best checkpoint. The logic for determining a new best metric was separated out into the `_determine_best_metric` method.
2. `_determine_best_metric` is called after every evaluation inside of `_maybe_log_save_evaluate`. The return value `new_best_metric` is used to determine whether or not a new best metric has been achieved, and if the save strategy is `"best"` then the TrainerControl's `should_save` flag is switched on.
    * Contrary to what I initially thought, `best_metric` does not seem to be tracked by default. Rather, it's only tracked when `args.metric_for_best_model` is provided. I believe that a best metric of some sort should always be tracked, and therefore if a value is not provided then the validation loss is used to determine a new best.
3. A new object called `SaveStrategy` was created in `trainer_utils` that adds a new attribute called `BEST` to the previous `IntervalStrategy`.

I'm not sure if I like the rather "hack-y" way that I implemented this by manually switching the TrainerControl's `should_save` flag rather than delegating it to the callback handler like the other flags are dealt with. The problem is that the flags are normally updated before calling `_maybe_log_save_evaluate` inside of the inner training loop, which means there's no way for us to determine whether or not a new best metric has been achieved with the current logic. I'm not sure if I'm making sense, but I'm open to any other suggestions.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

@muellerzr @SunMarc 

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @SunMarc

Integrations:

- deepspeed: HF Trainer/Accelerate: @muellerzr
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc 

Documentation: @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

seanswyi committed:
Added SaveStrategy and made according changes.

`save_strategy` previously followed `IntervalStrategy` but now follows
`SaveStrategy`.

Changes were made accordingly to the code and the docstring.

seanswyi committed:
Add _determine_best_metric and new saving logic.

1. Logic to determine the best logic was separated out from
`_save_checkpoint`.
2. In `_maybe_log_save_evaluate`, whether or not a new best metric was
achieved is determined after each evaluation, and if the save strategy
is "best' then the TrainerControl is updated accordingly.

seanswyi committed:
Added SaveStrategy.

Same as IntervalStrategy, but with a new attribute called BEST.

seanswyi committed:
IntervalStrategy -> SaveStrategy

seanswyi committed:
IntervalStratgy -> SaveStrategy for save_strat.

seanswyi committed:
Interval -> Save in docstring.

seanswyi committed:
Updated docstring for save_strategy.

seanswyi committed:
Changes from `make fixup`.

seanswyi committed:
Removed redundant metrics argument.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31817). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

amyeroberts commented:
Gentle ping @muellerzr @SunMarc

muellerzr commented on src/transformers/trainer.py:
```diff
@@ -2796,9 +2796,16 @@ def _maybe_log_save_evaluate(self, tr_loss, grad_norm, model, trial, epoch, igno
+            new_best_metric = self._determine_best_metric(metrics=metrics, trial=trial)
+
+            if self.args.save_strategy == SaveStrategy.BEST:
+                if new_best_metric:
+                    self.control.should_save = True
+                else:
+                    self.control_should_save = False
```
This can be refactored:

```suggestion
            self.control.should_save = new_best_metric
```

muellerzr commented on src/transformers/trainer.py:
```diff
@@ -2864,7 +2871,51 @@ def _load_rng_state(self, checkpoint):
+                    f"The available evaluation metrics are: {list(metrics.keys())}. Consider changing the `metric_for_best_model` via the TrainingArguments."
+                ) from exc
+
+            if self.args.greater_is_better:
+                operator = np.greater
+            else:
+                operator = np.less
```
Similarly here:

```suggestion
            operator = np.greater if self.args.greater_is_better else np.less
```

muellerzr reviewed (approved):
While this may seem hacky to you, I think this perfectly expands our given system. Any chance you could add a few tests to this? (Over in `test_trainer.py`). Very nice work!

seanswyi commented:
> While this may seem hacky to you, I think this perfectly expands our given system. Any chance you could add a few tests to this? (Over in `test_trainer.py`). Very nice work!

Thanks! And sure, I think I'll have some time to work on it over the weekend.

seanswyi committed:
Added new test_save_best_checkpoint test.

1. Checks for both cases where `metric_for_best_model` is explicitly
provided and when it's not provided.
2. The first case should have two checkpoints saved, whereas the second
should have three saved.

seanswyi committed:
Changed should_training_end saving logic.

The Trainer saves a checkpoints at the end of training by default as
long as `save_strategy != SaveStrategy.NO`. This condition was modified
to include `SaveStrategy.BEST` because it would be counterintuitive that
we'd only want the best checkpoint to be saved but the last one is as
well.

seanswyi committed:
`args.metric_for_best_model` default to loss.

seanswyi commented:
@muellerzr (cc. @SunMarc)

Hello. I've implemented a new test case inside of `tests/trainer/test_trainer.py` as suggested. Patches were used to control the output of the `_evaluate` method.

The first test case is when a value for `metric_for_best_model` is explicitly provided, and the second is when it's not provided and therefore defaults to the loss.

I've changed the DefaultFlowCallback object so that if `save_strategy` is either `"no"` or `"best"` then a checkpoint will not be saved at the end of training, as is the default behavior; it didn't really make sense to me that we'd want to keep only the best checkpoint but the last checkpoint is also being saved.

seanswyi committed:
Undo metric_for_best_model update.

seanswyi committed:
Remove checking metric_for_best_model.

seanswyi commented:
@muellerzr @SunMarc No rush, but just a gentle nudge/reminder! Thanks.

SunMarc commented on src/transformers/trainer.py:
```diff
@@ -2877,7 +2881,48 @@ def _load_rng_state(self, checkpoint):
+                ) from exc
+
+            operator = np.greater if self.args.greater_is_better else np.less
+        else:
+            metric_value = metrics["eval_loss"]
+            operator = np.less
+
+        if self.state.best_metric is None or operator(metric_value, self.state.best_metric):
+            run_dir = self._get_output_dir(trial=trial)
+            checkpoint_folder = f"{PREFIX_CHECKPOINT_DIR}-{self.state.global_step}"
+            output_dir = os.path.join(run_dir, checkpoint_folder)
+
+            self.state.best_metric = metric_value
+            self.state.best_model_checkpoint = output_dir
+
+            new_best_metric = True
+
+        return new_best_metric
```
I prefer that we don't set a hardcoded metric `eval_loss` because it will be hard to find the issue. Instead, we should raise an error if the user do not set `metric_for_best_model` when using `SaveStrategy.BEST` or `load_best_model_at_end`.Hence, the last if block should only happen when inside`if self.args.metric_for_best_model is not None:`. 
LMK if this sounds good to you !

seanswyi replied:
Thanks for the suggestion, that sounds good to me. Just to make sure I understood correctly:

1. Throw an error if the user selects `SaveStrategy.BEST` or `load_best_model_at_end` but did not set a `metric_for_best_model`.
2. Remove the else block with the hard-coding of `metric_value`, hence only setting `new_best_metric = True` when `self.args.metric_for_best_model is not None`.

Let me know if there are any problems with my understanding. Otherwise, I'll go ahead and align my branch with `main` and implement the changes!

SunMarc replied:
that's right. Sorry for the delay !

SunMarc reviewed (approved):
Thanks for the reminder @seanswyi ! I've added a suggestion. Please also fix the conflits and we should be good to merge !

seanswyi committed:
Added test cases for loss and no metric.

seanswyi committed:
Added error for metric and changed default best_metric.

seanswyi committed:
Removed unused import.

seanswyi commented on src/transformers/trainer.py:
```diff
@@ -3096,19 +3102,19 @@ def _determine_best_metric(self, metrics, trial):
-            run_dir = self._get_output_dir(trial=trial)
-            checkpoint_folder = f"{PREFIX_CHECKPOINT_DIR}-{self.state.global_step}"
-            output_dir = os.path.join(run_dir, checkpoint_folder)
+            if self.state.best_metric is None:
+                self.state.best_metric = float("-inf") if self.args.greater_is_better else float("inf")
+
+            if operator(metric_value, self.state.best_metric):
+                run_dir = self._get_output_dir(trial=trial)
+                checkpoint_folder = f"{PREFIX_CHECKPOINT_DIR}-{self.state.global_step}"
+                output_dir = os.path.join(run_dir, checkpoint_folder)
```
I changed the previous code of `if self.state.best_metric is None` to explicitly setting upper and lower bounds for the values for comparisons, along with separating it out from the value comparison (i.e., `operator`).

When the `metric_for_best_model` was set to a loss value, having the `self.state.best_metric` be `None` wasn't properly comparing the values.

Let me know if there'd be a better way.

ArthurZucker reviewed (commented):
Looks good, just a small nit!

seanswyi commented on src/transformers/trainer.py:
```diff
@@ -2988,9 +2994,13 @@ def _maybe_log_save_evaluate(self, tr_loss, grad_norm, model, trial, epoch, igno
         metrics = None
         if self.control.should_evaluate:
             metrics = self._evaluate(trial, ignore_keys_for_eval)
+            new_best_metric = self._determine_best_metric(metrics=metrics, trial=trial)
```
Thanks for the suggestion; I also prefer this since it's more descriptive of the variable. 👍

seanswyi committed:
`new_best_metric` -> `is_new_best_metric`

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

SunMarc reviewed (approved):
Thanks for iterating !

seanswyi committed:
Applied `is_new_best_metric` to all.

Changes were made for consistency and also to fix a potential bug.

muellerzr committed:
Merge branch 'main' into feat/new-best-save-strategy

ArthurZucker reviewed (approved):
Great work, thanks for iterating!

seanswyi commented:
@ArthurZucker @SunMarc @muellerzr Sorry to keep pinging you guys on this, but I'm noticing that the tests seem to be possibly stuck after the merge. Could anybody check this when they have the chance? Thanks.

https://github.com/huggingface/transformers/runs/32165247051

SunMarc commented:
Indeed, thanks for noticing @seanswyi but don't worry, I check the most recent [CI](https://github.com/huggingface/transformers/actions/runs/11657144637/job/32454510816) and the test is passing. I will just cancel the runs.

seanswyi commented:
> Indeed, thanks for noticing @seanswyi but don't worry, I check the most recent [CI](https://github.com/huggingface/transformers/actions/runs/11657144637/job/32454510816) and the test is passing. Can you cancel the run @ArthurZucker ? Or it will just cancel by itself after a few days ?

Ah, got it. Thanks for following up!

shcheklein commented on src/transformers/trainer.py:
```diff
@@ -419,6 +419,12 @@ def __init__(
             raise ValueError(
                 f"You have set `args.eval_strategy` to {args.eval_strategy} but you didn't pass an `eval_dataset` to `Trainer`. Either set `args.eval_strategy` to `no` or pass an `eval_dataset`. "
             )
+        if args.save_strategy == SaveStrategy.BEST or args.load_best_model_at_end:
```
@seanswyi Q: does this change mean that this logic:

https://huggingface.co/docs/transformers/main/main_classes/trainer#transformers.Seq2SeqTrainingArguments.metric_for_best_model

> ... Will default to "loss" if unspecified and load_best_model_at_end=True (to use the evaluation loss).

is broken now?

seanswyi replied:
I haven't tested it yet, but it seems like it is. The docs probably need an update.

seanswyi replied:
@shcheklein Yeah, I don't think that there's any case where `metric_for_best_model` default to loss.

umarbutler commented:
If `load_best_model_at_end = True`, the strategy cannot be set to `best`.

umarbutler commented:
This solves https://github.com/huggingface/transformers/issues/35070, cheers! 🥳
